Allow email privacy to be configured

sbose78 opened the pull request:
Fixes https://github.com/fabric8-services/fabric8-auth/issues/35

The objective of this PR is to provide a way to make email private using the `PATCH /api/users`

Make your email private : 
```
curl --request POST \
  --url http://localhost:8089/api/users \
  --header 'authorization: Bearer eyJhbGciOiJSUzI1NiIajg' \
  --header 'content-type: application/json' \
  --data '{"data": { "attributes": {"email_private":true},"type": "identities"}}'
```

The email should not be returned ( will be an empty string ) when a call similar to 
```
https://auth.openshift.io/api/users?filter[username]=shbose
```
is made.

However, a call to to 
```
curl --request GET \
  --url http://localhost:8089/api/user \
  --header 'authorization: Bearer eyJhbGciOiJSUzI1NiIajg' \
```

would return the email address irrespective of the `email_private` value.

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-1`

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-2`

alexeykazakov commented on migration/sql-files/012-hide-email.sql:
```diff
@@ -0,0 +1,3 @@
+-- Store identity provider username for the external token
+ALTER TABLE users ADD COLUMN email_hidden boolean;
```
email_private would be better. Hidden/shown email is more UI/UX thing. Private/public is a model thing.

alexeykazakov commented on controller/users.go:
```diff
@@ -82,7 +82,7 @@ func (c *UsersController) Show(ctx *app.ShowUsersContext) error {
 		}
 		return ctx.ConditionalRequest(*user, c.config.GetCacheControlUser, func() error {
-			return ctx.OK(ConvertToAppUser(ctx.RequestData, user, identity))
+			return ctx.OK(ConvertToAppUser(ctx.RequestData, user, identity, true))
```
should be false, right?

sbose78 replied:
nope, should be a `true` since this is the `GET /api/user` with `Authorization..`

sbose78 replied:
My bad, this is the `/api/users/<ID>` .

alexeykazakov commented on controller/users.go:
```diff
@@ -863,7 +868,7 @@ func loadKeyCloakIdentity(appl application.Application, user account.User) (*acc
 
 // ConvertToAppUser converts a complete Identity object into REST representation
-func ConvertToAppUser(request *goa.RequestData, user *account.User, identity *account.Identity) *app.User {
+func ConvertToAppUser(request *goa.RequestData, user *account.User, identity *account.Identity, isAuthenticated bool) *app.User {
```
Add the attribute description to the function comment what isAuthenticated means please.

sbose78 replied:
thanks 287005aa5744cef0eeea2c44c3c42b6f462d2d97

alexeykazakov commented on design/account.go:
```diff
@@ -203,6 +203,7 @@ var userDataAttributes = a.Type("UserDataAttributes", func() {
 	a.Attribute("username", d.String, "The username")
 	a.Attribute("registrationCompleted", d.Boolean, "Whether the registration has been completed")
 	a.Attribute("email", d.String, "The email")
+	a.Attribute("email_hidden", d.Boolean, "Whether the email address would be visible to other users.")
```
email_private

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-3`

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-4`

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-5`

kishansagathiya commented on design/account.go:
```diff
@@ -203,6 +203,7 @@ var userDataAttributes = a.Type("UserDataAttributes", func() {
 	a.Attribute("username", d.String, "The username")
 	a.Attribute("registrationCompleted", d.Boolean, "Whether the registration has been completed")
 	a.Attribute("email", d.String, "The email")
+	a.Attribute("email_private", d.Boolean, "Whether the email address would be visible to other users.")
```
email_private=true means not visible,
email_private=false means visible, right?
With that logic the proper description should be "Whether the email address wouldn't be visible to other users" or  "Whether the email address would be private"

kishansagathiya commented on design/account.go:
```diff
@@ -221,6 +222,7 @@ var updateUserDataAttributes = a.Type("UpdateIdentityDataAttributes", func() {
 	a.Attribute("email", d.String, "The email")
 	a.Attribute("bio", d.String, "The bio")
 	a.Attribute("url", d.String, "The url")
+	a.Attribute("email_private", d.Boolean, "Whether the email address would be visible to other users.")
```
Same as earlier,
"Whether the email address wouldn't be visible to other users" or "Whether the email address would be private"

kishansagathiya commented on account/user.go:
```diff
@@ -23,8 +23,9 @@ import (
-	Email              string             `sql:"unique_index"`                                            // This is the unique email field
+	ID                 uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	Email              string    `sql:"unique_index"`
+	EmailPrivate       bool
```
Since it is a boolean instead of EmailPrivate, IsEmailPrivate would make more sense. But it might involve quite a bit of refactoring, so you can ignore this.

sbose78 replied:
I agree with you - but in most places we aren't using the term `is..`  - hence will keep it this way for consistency. :)

alexeykazakov replied:
I don't think we should use "Is" in bool field names at all.

sbose78 commented:
@kishansagathiya Addressed your comments and updated description too - if you could test and give me a Green signal, I will proceed with merging this :)

alexeykazakov commented on controller/users_blackbox_test.go:
```diff
@@ -804,6 +804,53 @@ func (s *TestUsersSuite) TestListUsersByEmailOKEmptyResult() {
+	require.Equal(s.T(), "", *returnedUserResult.Attributes.Email)
+
+	// even though the email_private=true,
+	// the email address is visible to the user if her user token is passed.
```
The comment and test do not match. This is not what you are testing here. You are testing that api/users/:identityID doesn't return private emails. We should keep this test but fix the comment and add another test to the user controller test to actually test what your describe in the comment :)

sbose78 replied:
Agreed, looks like the /api/user and the /api/users confused me.
Thanks!

6583a918d37a389ab2aba36e546e01c89965f419

alexeykazakov commented on controller/users_blackbox_test.go:
```diff
@@ -804,6 +804,53 @@ func (s *TestUsersSuite) TestListUsersByEmailOKEmptyResult() {
+	email := user1.Email
+
+	// by default, email is public.
+	_, result := test.ListUsersOK(s.T(), nil, nil, s.controller, &email, nil, nil, nil)
```
Also check if test.ShowUsersOK() returns the public email.

sbose78 replied:
6583a918d37a389ab2aba36e546e01c89965f419

alexeykazakov commented on controller/users_blackbox_test.go:
```diff
@@ -822,12 +869,13 @@ func (s *TestUsersSuite) TestListUsersByEmailNotModifiedUsingIfNoneMatchHeader()
 
 func (s *TestUsersSuite) createRandomUser(fullname string) account.User {
```
Maybe we can replace this function by test.CreateTestUser() too?

sbose78 replied:
For consistency's sake, I will keep it as it is - because if change here, it *should* be changed in all other places in this file. 
In fact in this PR, https://github.com/fabric8-services/fabric8-auth/pull/227/files#diff-cc740fdbd6f22db6da588050e1b7d5a4R78
I have included a util method there too :)

alexeykazakov replied:
So? Is this a huge refactoring? :) I mean we should try to keep our tests clean. And try to improve our tests when it makes sense as well as we improve our main code. And if you already changing these tests it's fine to do some moderate test refactoring of the related code here. But it's also OK to do the test refactoring in the separate PR but I don't want us to create report more technical debt issues and forget about them ;)

sbose78 replied:
Cool, I will make refactor in this PR :)

sbose78 replied:
done - https://github.com/fabric8-services/fabric8-auth/pull/237/commits/f826965bb991df019f209ecb399ab385edbfab4e

alexeykazakov commented on migration/sql-files/012-hide-email.sql:
```diff
@@ -0,0 +1,3 @@
+-- Store identity provider username for the external token
+ALTER TABLE users ADD COLUMN email_private boolean;
+UPDATE USERS SET email_private = false;
```
Maybe set it to NOT_NULL after updating the table?

sbose78 replied:
7c096653afb69415af9e1ad559255cddbfba6e04

sbose78 commented on controller/search_user_blackbox_test.go:
```diff
@@ -140,6 +140,70 @@ func (s *TestSearchUserSearch) createTestData() []account.Identity {
+
+func (s *TestSearchUserSearch) TestEmailNotPrivateSearchOK() {
+	randomName := uuid.NewV4().String()
+	application.Transactional(s.Application, func(app application.Application) error {
```
2f43cfbac1c8050cc6aa6952a26b04d62d7aab9a

sbose78 commented on controller/search_user_blackbox_test.go:
```diff
@@ -140,6 +140,70 @@ func (s *TestSearchUserSearch) createTestData() []account.Identity {
 
+func (s *TestSearchUserSearch) TestEmailPrivateSearchOK() {
+	randomName := uuid.NewV4().String()
+	application.Transactional(s.Application, func(app application.Application) error {
```
2f43cfbac1c8050cc6aa6952a26b04d62d7aab9a

sbose78 commented:
All review comments taken care of. Thanks @alexeykazakov !

alexeykazakov commented:
This PR will break email notifications for users with private email.
What if we check if there is a service account token passed to /api/users and /api/space/:spaceID/collaborators endpoints and then include private emails to the result? WDYT? Then notification service could use it's service account to get users.

alexeykazakov commented:
We could address the email notification for private emails in a separate issue/PR though.

sbose78 commented:
Agreed, was wondering the same. 
Here's what we could do

1. When notification service comes up, call Auth service API to request for service account token, and keep it in memory.

2. Subsequently use the token from configuration memory object.

3. On the auth side, check for service account token, and based on that show the email .

How does this sound to you?

alexeykazakov commented:
> How does this sound to you?

Sounds good to me.

fabric8cd commented:
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-237-9`

kishansagathiya commented on controller/users_blackbox_test.go:
```diff
@@ -72,7 +72,9 @@ func (s *TestUsersSuite) SecuredServiceAccountController(identity account.Identi
 	// given
 	user := s.createRandomUser("TestUpdateUserOK")
-	identity := s.createRandomIdentity(user, account.KeycloakIDP)
+	identity, err := testsupport.CreateTestUser(s.DB, &user)
```
Isn't this creating user twice?

sbose78 replied:
`createRandomUser` only creates the user object ( but not in db ).
`CreateTestUser` creates the user in the db. 

( we should refactor the method name maybe )

kishansagathiya replied:
Cool , i am gonna rename this to `giveUser`. There are other files which has the same method, but that saves user in the database as well
